File: packages/gaia-explorer/src/gaia_explorer/validation.py
"""Validation helpers for GaiaExplorer JSON outputs."""
from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
from typing import Dict, Iterable, List, Optional, Tuple

import json

DEFAULT_JSON_PATH = Path(__file__).resolve().parents[2] / "out" / "gaia_grid.json"

NATURE_TYPES = {"main-sequence", "white-dwarf", "neutron-star", "black-hole"}
TEMP_MIN = 800.0
TEMP_MAX = 40000.0
MAIN_SEQUENCE_TEMP_MIN = 900.0
MAIN_SEQUENCE_TEMP_MAX = 10000.0
WHITE_DWARF_TEMP_MIN = 2500.0
WHITE_DWARF_TEMP_MAX = 40000.0


@dataclass(slots=True)
class ValidationResult:
    success: bool
    issues: List[str]

    def __bool__(self) -> bool:  # pragma: no cover - convenience
        return self.success


def validate_file(path: Path) -> ValidationResult:
    data = json.loads(path.read_text())
    issues: List[str] = []
    temperatures: List[float] = []
    nature_counts: Dict[str, int] = {}

    for cube_id, cube in data.get("cubes", {}).items():
        for star in cube.get("stars", []):
            name = star.get("name", "<unknown>")
            nature = star.get("nature")
            temp = star.get("temperature")

            if nature is None:
                issues.append(f"Missing nature for {name} in cube {cube_id}")
            elif nature not in NATURE_TYPES:
                issues.append(f"Unexpected nature '{nature}' for {name} in cube {cube_id}")
            else:
                nature_counts[nature] = nature_counts.get(nature, 0) + 1

            if temp is None:
                issues.append(f"Missing temperature for {name} in cube {cube_id}")
            else:
                temperatures.append(temp)
                if temp < TEMP_MIN or temp > TEMP_MAX:
                    issues.append(f"Temperature {temp}K out of bounds for {name} in cube {cube_id}")

                if nature == "main-sequence" and not (MAIN_SEQUENCE_TEMP_MIN <= temp <= MAIN_SEQUENCE_TEMP_MAX):
                    issues.append(
                        f"Main-sequence star {name} has suspicious temperature {temp}K in cube {cube_id}"
                    )
                if nature == "white-dwarf" and temp < WHITE_DWARF_TEMP_MIN:
                    issues.append(f"White dwarf {name} has low temperature {temp}K in cube {cube_id}")

    if not data.get("metadata"):
        issues.append("Missing metadata block")

    success = not issues
    return ValidationResult(success=success, issues=issues)


def validate_default_output() -> ValidationResult:
    return validate_file(DEFAULT_JSON_PATH)


def main() -> None:  # pragma: no cover - CLI entry point
    result = validate_default_output()
    if result.success:
        print("Validation passed.")
    else:
        print("Validation failed:")
        for issue in result.issues:
            print(f" - {issue}")
        raise SystemExit(1)


if __name__ == "__main__":  # pragma: no cover - CLI entry point
    main()

File: packages/gaia-explorer/src/validation.ts
import { readFile } from "node:fs/promises";

import { z } from "zod";

export const TemperatureMin = 800;
export const TemperatureMax = 40000;
export const MainSequenceTemperatureMin = 900;
export const MainSequenceTemperatureMax = 10000;
export const WhiteDwarfTemperatureMin = 2500;
const StarSchema = z.object({
    name: z.string(),
    relative_position: z.tuple([z.number(), z.number(), z.number()]),
    temperature: z.number().nullable(),
    nature: z.enum(["main-sequence", "white-dwarf", "neutron-star", "black-hole"]).nullable(),
});
export const DatasetSchema = z.object({
    metadata: z.record(z.string(), z.unknown()),
    selection: z.record(z.string(), z.unknown()).optional(),
    cubes: z.record(
        z.string(),
        z.object({
            index: z.tuple([z.number(), z.number(), z.number()]),
            origin: z.tuple([z.number(), z.number(), z.number()]),
            stars: z.array(StarSchema),
        }),
    ),
});
export type ValidationResult = Readonly<{
    success: boolean;
    schemaIssues: ReadonlyArray<string>;
    issues: ReadonlyArray<string>;
}>;
type ValidatedStar = z.infer<typeof StarSchema>;
function validateStar(star: ValidatedStar, cubeId: string): ReadonlyArray<string> {
    const issues: string[] = [];
    if (star.nature === null) {
        issues.push(`Missing nature for ${star.name} in cube ${cubeId}`);
    }
    if (star.temperature === null) {
        issues.push(`Missing temperature for ${star.name} in cube ${cubeId}`);
        return issues;
    }
    if (star.temperature < TemperatureMin || star.temperature > TemperatureMax) {
        issues.push(`Temperature ${star.temperature}K out of bounds for ${star.name} in cube ${cubeId}`);
    }
    if (
        star.nature === "main-sequence" &&
        (star.temperature < MainSequenceTemperatureMin || star.temperature > MainSequenceTemperatureMax)
    ) {
        issues.push(
            `Main-sequence star ${star.name} has suspicious temperature ${star.temperature}K in cube ${cubeId}`,
        );
    }
    if (star.nature === "white-dwarf" && star.temperature < WhiteDwarfTemperatureMin) {
        issues.push(`White dwarf ${star.name} has low temperature ${star.temperature}K in cube ${cubeId}`);
    }
    return issues;
}
export function validateDataset(input: unknown): ValidationResult {
    const parsed = DatasetSchema.safeParse(input);
    if (!parsed.success) {
        return {
            success: false,
            schemaIssues: parsed.error.issues.map((issue) => `${issue.path.join(".")}: ${issue.message}`),
            issues: [],
        };
    }
    const issues = Object.entries(parsed.data.cubes).flatMap(([cubeId, cube]) =>
        cube.stars.flatMap((star) => validateStar(star, cubeId)),
    );
    if (Object.keys(parsed.data.metadata).length === 0) {
        issues.push("Missing metadata block");
    }
    return { success: issues.length === 0, schemaIssues: [], issues };
}
export async function validateFile(path: string): Promise<ValidationResult> {
    return validateDataset(JSON.parse(await readFile(path, "utf8")) as unknown);
}
